solidity/test/maintainer/MaintainerProxy.test.ts

```diff
@@ -0,0 +1,3434 @@
+        ).to.be.revertedWith("Ownable: caller is not the owner")
+      })
+    })
+
+    context("when the caller is the owner", () => {
+      let tx: ContractTransaction
+
+      before(async () => {
+        await createSnapshot()
+        const maintainers = await maintainerProxy.getAllMaintainers()
+        console.log("maintainers.length", maintainers[0])
+
+        tx = await maintainerProxy
+          .connect(governance)
+          .authorize(thirdParty.address)
+      })
+
+      after(async () => {
+        await restoreSnapshot()
+      })
+
+      it("should be already populated with the authorized maintainer", async () => {
+        expect(
+          await maintainerProxy.isAuthorized(authorizedMaintainer.address)
+        ).to.be.equal(1)
+      })
+
+      it("should authorize a thirdParty", async () => {
+        expect(
+          await maintainerProxy.isAuthorized(thirdParty.address)
+        ).to.be.equal(2)
+      })
+
+      it("should be total of 2 authorized maintainers", async () => {
+        const maintainers = await maintainerProxy.getAllMaintainers()
+        expect(maintainers.length).to.be.equal(2)
+      })
+
+      it("should add a thirdParty to a maintainers list", async () => {
+        const contract = await maintainerProxy.maintainers(1)
+        expect(contract).to.be.equal(thirdParty.address)
+      })
+
+      it("should emit a MaintainerAuthorized event", async () => {
+        await expect(tx)
+          .to.emit(maintainerProxy, "MaintainerAuthorized")
+          .withArgs(thirdParty.address)
+      })
+    })
+  })
+
+  describe("unauthorize", () => {
+    before(async () => {
+      await createSnapshot()
+
+      await maintainerProxy
+        .connect(governance)
+        .unauthorize(authorizedMaintainer.address)
+    })
+
+    after(async () => {
+      await restoreSnapshot()
+    })
+
+    context("when the caller is not the owner", () => {
+      it("should revert", async () => {
+        await expect(
+          maintainerProxy.connect(thirdParty).unauthorize(thirdParty.address)
+        ).to.be.revertedWith("Ownable: caller is not the owner")
+      })
+    })
+
+    context("when the caller is the owner", () => {
+      it("should be a total of 0 authorized maintainers", async () => {
+        const authorizedMaintainers = await maintainerProxy.getAllMaintainers()
+        await expect(authorizedMaintainers.length).to.be.equal(0)
+      })
+
+      context("when there are no authorized maintainers", () => {
+        before(async () => {
+          await createSnapshot()
+        })
+
+        after(async () => {
+          await restoreSnapshot()
+        })
+
+        it("should revert", async () => {
+          await expect(
+            maintainerProxy.connect(governance).unauthorize(thirdParty.address)
+          ).to.be.revertedWith("No maintainer to unauthorize")
+        })
+      })
+
+      context("when there are authorized maintainers", () => {
```

Comment:
Should we remove this context and move the authorization of unnamed signers to the main setup of unauthorize context?